contracts/metatx/ERC2771Forwarder.sol

```diff
@@ -0,0 +1,240 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts (last updated v4.9.0) (metatx/ERC2771Forwarder.sol)
+
+pragma solidity ^0.8.19;
+
+import "../utils/cryptography/ECDSA.sol";
+import "../utils/cryptography/EIP712.sol";
+import "../utils/Nonces.sol";
+
+/**
+ * @dev An implementation of a production-ready forwarder compatible with ERC2771 contracts.
+ *
+ * This forwarder operates on forward requests that include:
+ *
+ * * `from`: An address to operate on behalf of. It is required to be equal to the request signer.
+ * * `to`: The address that should be called.
+ * * `value`: The amount of ETH to attach with the requested call.
+ * * `gas`: The amount of gas limit that will be forwarded with the requested call.
+ * * `nonce`: A unique transaction ordering identifier to avoid replayability and request invalidation.
+ * * `deadline`: A timestamp after which the request is not executable anymore.
+ * * `data`: Encoded `msg.data` to send with the requested call.
+ */
+contract ERC2771Forwarder is EIP712, Nonces {
+    using ECDSA for bytes32;
+
+    struct ForwardRequest {
+        address from;
+        address to;
+        uint256 value;
+        uint256 gas;
+        uint256 nonce;
+        uint48 deadline;
+        bytes data;
+    }
+
+    bytes32 private constant _FORWARD_REQUEST_TYPEHASH =
+        keccak256(
+            "ForwardRequest(address from,address to,uint256 value,uint256 gas,uint256 nonce,uint48 deadline,bytes data)"
```

Comment:
I believe we agreed to add `address relayer` here. My proposal is to add this field, if it has a zero address anyone is allowed to relay it (include it in a `forward` function call), and if it has a non-zero address then it must be equal to the relayer (`msg.sender`).

Note that it doesn't need to be included in the struct, because we already have `msg.sender` available, but this forces us to try both `msg.sender` and `address(0)` as possible values, which will incur some additional cost so we need to make sure to implement it efficiently.

Reply:
After discussing further, we agree to disregard the frontrunning (which only becomes an issue if relayers are incentivized -- not the case here), and not add the relayer parameter to forward requests.